test/distributed/test_c10d_spawn_nccl.py

```diff
@@ -64,8 +83,170 @@ def test_shared_allgather_nccl(self):
                 ProcessGroupShareTensorTest._test_allgather_process,
                 [torch.ones(2, 2).to(i) * i for i in range(self.world_size)],
                 ProcessGroupShareTensorTest._init_pg_nccl,
-                self.world_size)
+                self.world_size,
+            )
+
+
+# Skip dev-asan as torch + multiprocessing spawn have known issues
+if not TEST_WITH_DEV_DBG_ASAN:
+
+    class TestDistributedNNFunctions(MultiProcessTestCase):
+        def setUp(self):
+            super(TestDistributedNNFunctions, self).setUp()
+            self._spawn_processes()
+
+        def tearDown(self):
+            super(TestDistributedNNFunctions, self).tearDown()
+            try:
+                os.remove(self.file_name)
+            except OSError:
+                pass
+
+        @property
+        def op_timeout_sec(self):
+            return 1
+
+        @property
+        def world_size(self):
+            return 2
+
+        @requires_nccl()
+        @skip_if_lt_x_gpu(2)
+        @sandcastle_skip_if(
+            not _torch_dist_nn_available, "torch.distributed.nn is not available"
+        )
+        def test_broadcast(self):
+            store = c10d.FileStore(self.file_name, self.world_size)
+            # This is required because these functions calls directly to the .dist and needs
+            # the world to be initialized
+            c10d.init_process_group(
+                store=store, rank=self.rank, world_size=self.world_size, backend="nccl"
+            )
+            device = torch.device(f"cuda:{self.rank}")
+            x = torch.ones(5, 5, device=device) + self.rank
+            x.requires_grad = True
+            y = torch.distributed.nn.broadcast(x, 1)
+            self.assertEqual(y, 1 + torch.ones(5, 5))
+            z = y.sin().sum()
+            z.backward()
+            # We can't check the gradient of communications numerically so we have to do some calculations
+            if self.rank == 1:
+                self.assertEqual(x.grad, 2 * torch.cos(x))
+            elif self.rank == 0:
+                self.assertEqual(x.grad, torch.zeros(5, 5, device=device))
```

Comment:
Can we reuse this code between `test_c10d_spawn_gloo` and `test_c10d_spawn_nccl` instead of repeating it here? There is `test_c10d_common.py` where we can probably put the common logic.

Reply:
Sure, let me refactor it a little bit and move the commonly supported module to the common.